third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -0,0 +1,191 @@
+  ALL_DS = 0x00000080,
+  DS_READ = 0x00000100,
+  DS_WRITE = 0x00000200
+};
+
+// Create an intrinsic to control how different instruction kinds should
+// interleave for better ILP.
+void createSchedGroupBarrier(PatternRewriter &rewriter,
+                             InstructionKindMask maskValue, int sizeValue,
+                             int groupIdValue) {
+  MLIRContext *ctx = rewriter.getContext();
+  Location loc = rewriter.getUnknownLoc();
+  auto intrinsicName = StringAttr::get(ctx, "llvm.amdgcn.sched.group.barrier");
+  LLVM::FastmathFlagsAttr defaultFlags{};
+  Type i32 = rewriter.getI32Type();
+  auto mask = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, static_cast<int64_t>(maskValue)));
+  auto size = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, sizeValue));
+  auto groupId = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, groupIdValue));
+  rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
+                                         ValueRange{mask, size, groupId},
+                                         defaultFlags);
+}
+
+// Insert intrinsic that controls the types of instructions that may be
+// allowed to cross the intrinsic during instruction scheduling
+Operation *createSchedBarrier(PatternRewriter &rewriter, int64_t maskValue) {
+  MLIRContext *ctx = rewriter.getContext();
+  Location loc = rewriter.getUnknownLoc();
+  auto intrinsicName = StringAttr::get(ctx, "llvm.amdgcn.sched.barrier");
+  LLVM::FastmathFlagsAttr defaultFlags{};
+  Type i32 = rewriter.getI32Type();
+  auto mask = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, maskValue));
+  return rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
+                                                ValueRange{mask}, defaultFlags);
+}
+
+// Insert an experimental intrinsic for instruction group level parallelism.
+// The intrinsic takes a value that specifies the strategy.
+Operation *createIglpOpt(PatternRewriter &rewriter, int value) {
+  MLIRContext *ctx = rewriter.getContext();
+  Location loc = rewriter.getUnknownLoc();
+  auto intrinsicName = StringAttr::get(ctx, "llvm.amdgcn.iglp.opt");
+  LLVM::FastmathFlagsAttr defaultFlags{};
+  Type i32 = rewriter.getI32Type();
+  auto mask = rewriter.create<LLVM::ConstantOp>(
+      loc, rewriter.getIntegerAttr(i32, static_cast<int64_t>(value)));
+  return rewriter.create<LLVM::CallIntrinsicOp>(loc, TypeRange{}, intrinsicName,
+                                                ValueRange{mask}, defaultFlags);
+}
+
+struct InstructionSchedHintsRewriter
+    : public OpRewritePattern<triton::amdgpu::InstructionSchedHint> {
+
+  InstructionSchedHintsRewriter(mlir::MLIRContext *ctx, uint32_t variant)
+      : OpRewritePattern(ctx) {
+    schedulingType = static_cast<SchedulingType>(variant);
+    schedulingType = schedulingType > SchedulingType::COUNT
```

Comment:
It's better to explicitly error out to avoid suprises rather than "magically make it work".

Reply:
Done. Added the assert statement